4 - AnimalShelter/Code/AnimalShelter.cs

```diff
@@ -1,55 +1,61 @@
-﻿using AnimalShelterCode;
-using System;
+﻿using System;
 using System.Collections.Generic;
-using System.Text;
 using System.Linq;
 
-namespace AnimalShelterCode
+namespace AnimalShelters.Code
 {
     public class AnimalShelter
     {
-        public int AnimalCounter { get; private set; }
-
-        public List<Animal> Animals { get; private set; }
-
+        public List<Animal> Animals { get { return _Animals; } }
         public AnimalShelter()
         {
-            Animals = new List<Animal>();
-            AnimalCounter = 1;
+            _Animals = new List<Animal>();
         }
 
         public OperationResult AddAnimal(Animal animal)
         {
-            if (animal.Type == AnimalType.Bear || animal.Type == AnimalType.Deer) 
-            { 
-                return new OperationResult($"Animal Type {animal.Type} is not Supported");
+            if (IsSupported(animal.Type)) 
+            {
+                animal.AssignId();
+                _Animals.Add(animal);
+                return new OperationResult(animal);
             }
             else
             {
-                animal.UniqueAnimalId = AnimalCounter++;
-                Animals.Add(animal);
-                return new OperationResult(animal);
+                return new OperationResult($"Animal Type {animal.Type} is not Supported");
             }
         }
 
-        public List<Animal> GetAnimals(AnimalType filtertype)
+        public List<Animal> GetAnimals(FilterObject filter)
         {
-            var filteredlist = from animal in Animals
-                               where animal.Type == filtertype
-                               select animal;
+            var filteredList = _Animals;
 
-            return filteredlist.ToList();
+            if( filter.Type.HasValue && filter.CanFly.HasValue)
+            {
+                filteredList = _Animals.Where(a => a.CanFly == filter.CanFly
+                                             && a.Type == filter.Type).ToList();
+            }
+            else if (filter.Type.HasValue)
+            {
+                filteredList = _Animals.Where(a => a.Type == filter.Type).ToList();
+            }
+            else if (filter.CanFly.HasValue)
+            {
+                filteredList = _Animals.Where(a => a.CanFly == filter.CanFly).ToList();
+            }
+
+            return filteredList;
         }
         public List<Animal> GetAnimals()
         {
-            return Animals;
+            return _Animals;
```

Comment:
I would advise returning _animals.ToList(). The reason for the ToList() is that returns a *clone* of the list, which means if they add or remove to that list it won't change yours.